napari/components/viewer_model.py

```diff
@@ -282,22 +279,29 @@ def interactive(self, interactive):
     def active_layer(self):
         """int: index of active_layer
         """
-        return self._active_layer
+        warnings.warn(
+            (
+                "The viewer.active_layer property is deprecated and will be removed after version 0.4.5."
+                " Instead you should use the viewer.layers.active property"
+            ),
+            category=DeprecationWarning,
+            stacklevel=2,
+        )
+        return self.layers.active
 
     @active_layer.setter
     def active_layer(self, active_layer):
-        if active_layer == self.active_layer:
+        warnings.warn(
+            (
+                "The viewer.active_layer property is deprecated and will be removed after version 0.4.5."
+                " Instead you should use the viewer.layers.active property"
+            ),
+            category=DeprecationWarning,
+            stacklevel=2,
+        )
+        if active_layer == self.layers.active:
             return
-
-        if self._active_layer is not None:
-            self.keymap_providers.remove(self._active_layer)
-
-        self._active_layer = active_layer
-
-        if active_layer is not None:
-            self.keymap_providers.insert(0, active_layer)
-
-        self.events.active_layer(item=self._active_layer)
+        self.layers.active = active_layer
```

Comment:
Triple iteration over the list of active layers. 
I think that this `if` protection check should be hidden in `self.layers.active` (which case cascade of events every time the active layer is set, even if it is already active one).

Reply:
I'll change this to use cache